### PR TITLE
Fix  backfill  queued  task getting reset to scheduled state. 

### LINKED ISSUE #23145: Task stuck in "scheduled" when running in backfill job
### Apache Airflow version

2.2.4

### What happened

We are running airflow 2.2.4 with KubernetesExecutor. I have created a dag to run airflow backfill command with SubprocessHook. What was observed is that when I started to backfill a few days' dagruns the backfill would get stuck with some dag runs having tasks staying in the "scheduled" state and never getting running. 

We are using the default pool and the pool is totoally free when the tasks got stuck.

I could find some logs saying:
`TaskInstance: <TaskInstance: test_dag_2.task_1 backfill__2022-03-29T00:00:00+00:00 [queued]> found in queued state but was not launched, rescheduling` and nothing else in the log.

### What you think should happen instead

The tasks stuck in "scheduled" should start running when there is free slot in the pool.

### How to reproduce

Airflow 2.2.4 with python 3.8.13, KubernetesExecutor running in AWS EKS.

One backfill command example is: `airflow dags backfill test_dag_2 -s 2022-03-01 -e 2022-03-10 --rerun-failed-tasks`

The test_dag_2 dag is like:

```
import time
from datetime import timedelta

import pendulum
from airflow import DAG
from airflow.decorators import task
from airflow.models.dag import dag
from airflow.operators.bash import BashOperator
from airflow.operators.dummy import DummyOperator
from airflow.operators.python import PythonOperator

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'email': ['airflow@example.com'],
    'email_on_failure': True,
    'email_on_retry': False,
    'retries': 1,
    'retry_delay': timedelta(minutes=5),
}


def get_execution_date(**kwargs):
    ds = kwargs['ds']
    print(ds)

with DAG(
        'test_dag_2',
        default_args=default_args,
        description='Testing dag',
        start_date=pendulum.datetime(2022, 4, 2, tz='UTC'),
        schedule_interval="@daily", catchup=True, max_active_runs=1,
) as dag:
    t1 = BashOperator(
        task_id='task_1',
        depends_on_past=False,
        bash_command='sleep 30'
    )

    t2 = PythonOperator(
        task_id='get_execution_date',
        python_callable=get_execution_date
    )

    t1 >> t2
```

### Operating System

Debian GNU/Linux

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==3.0.0
apache-airflow-providers-celery==2.1.0
apache-airflow-providers-cncf-kubernetes==3.0.2
apache-airflow-providers-docker==2.4.1
apache-airflow-providers-elasticsearch==2.2.0
apache-airflow-providers-ftp==2.0.1
apache-airflow-providers-google==6.4.0
apache-airflow-providers-grpc==2.0.1
apache-airflow-providers-hashicorp==2.1.1
apache-airflow-providers-http==2.0.3
apache-airflow-providers-imap==2.2.0
apache-airflow-providers-microsoft-azure==3.6.0
apache-airflow-providers-microsoft-mssql==2.1.0
apache-airflow-providers-odbc==2.0.1
apache-airflow-providers-postgres==3.0.0
apache-airflow-providers-redis==2.0.1
apache-airflow-providers-sendgrid==2.0.1
apache-airflow-providers-sftp==2.4.1
apache-airflow-providers-slack==4.2.0
apache-airflow-providers-snowflake==2.5.0
apache-airflow-providers-sqlite==2.1.0
apache-airflow-providers-ssh==2.4.0

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #23356: Tasks set to queued by a backfill get cleared and rescheduled by the kubernetes executor, breaking the backfill
### Apache Airflow version

2.2.5 (latest released)

### What happened

A backfill launched from the scheduler pod, queues tasks as it should but while they are in the process of starting the kubernentes executor loop running in the scheduler clears these tasks and reschedules them via this function https://github.com/apache/airflow/blob/9449a107f092f2f6cfa9c8bbcf5fd62fadfa01be/airflow/executors/kubernetes_executor.py#L444

This causes the backfill to not queue any more tasks and enters an endless loop of waiting for the task it has queued to complete.

The way I have mitigated this is to set the `AIRFLOW__KUBERNETES__WORKER_PODS_QUEUED_CHECK_INTERVAL` to 3600, which is not ideal

### What you think should happen instead

The function clear_not_launched_queued_tasks should respect tasks launched by a backfill process and not clear them. 

### How to reproduce

start a backfill with large number of tasks and watch as they get queued and then subsequently rescheduled by the kubernetes executor running in the scheduler pod

### Operating System

Debian GNU/Linux 10 (buster)

### Versions of Apache Airflow Providers

```

apache-airflow            2.2.5            py38h578d9bd_0    
apache-airflow-providers-cncf-kubernetes 3.0.2              pyhd8ed1ab_0    
apache-airflow-providers-docker 2.4.1              pyhd8ed1ab_0    
apache-airflow-providers-ftp 2.1.2              pyhd8ed1ab_0    
apache-airflow-providers-http 2.1.2              pyhd8ed1ab_0    
apache-airflow-providers-imap 2.2.3              pyhd8ed1ab_0    
apache-airflow-providers-postgres 3.0.0              pyhd8ed1ab_0    
apache-airflow-providers-sqlite 2.1.3              pyhd8ed1ab_0    

```

### Deployment

Other 3rd-party Helm chart

### Deployment details

Deployment is running the latest helm chart of Airflow Community Edition

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #23693: Airflow backfill command not working; Tasks are stuck in scheduled forever.
### Apache Airflow version

2.2.1

### What happened

When I try to backfill using `airflow dags backfill ...`, the 'Run' shows running but the tasks are stuck in scheduled state forever.

1. Code for DAG to reproduce the problem

```python
import time
from datetime import timedelta

import pendulum
from airflow import DAG
from airflow.decorators import task
from airflow.models.dag import dag
from airflow.operators.bash import BashOperator
from airflow.operators.dummy import DummyOperator
from airflow.operators.python import PythonOperator

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'email': ['airflow@example.com'],
    'email_on_failure': True,
    'email_on_retry': False,
    'retries': 2,
    'retry_delay': timedelta(minutes=5),
}


def get_execution_date(**kwargs):
    ds = kwargs['ds']
    print(ds)

with DAG(
        'test_dag_2',
        default_args=default_args,
        description='Testing dag',
        start_date=pendulum.datetime(2022, 4, 2, tz='UTC'),
        schedule_interval="@hourly", max_active_runs=5, concurrency=10,
) as dag:
    t1 = BashOperator(
        task_id='task_1',
        depends_on_past=False,
        bash_command='sleep 5'
    )

    t2 = PythonOperator(
        task_id='get_execution_date',
        python_callable=get_execution_date
    )

    t1 >> t2
```

2. Airflow backfill CLI

```bash
$ airflow dags backfill --subdir /opt/airflow/dags/repo --reset-dagruns -s "2022-03-01 01:00:00" -e "2022-03-02 01:00:00" test_dag_2
```

3. Result
image

4. Discussion
The 'Run' shows running but the tasks are stuck in scheduled state forever.
Here are some variants what I have checked. (but still have this issue.)
- max_active_runs, concurrency
- -x option for backfill command
- --subdir option for backfill command
- --reset-dagruns for backfill command
- schedule_interval to `None` and `@once`
- Restart the Airflow scheduler pods.
- **Upgrade Airflow version to 2.3.0.**

5. Related issue
https://github.com/apache/airflow/issues/13542

### What you think should happen instead

The backfill jobs should successfully done from the past.

### How to reproduce

_No response_

### Operating System

CentOS Linux release 7.9.2009 (Core)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon          3.3.0
apache-airflow-providers-celery          2.1.0
apache-airflow-providers-cncf-kubernetes 3.0.2
apache-airflow-providers-docker          2.6.0
apache-airflow-providers-elasticsearch   3.0.3
apache-airflow-providers-ftp             2.1.2
apache-airflow-providers-grpc            2.0.4
apache-airflow-providers-hashicorp       2.2.0
apache-airflow-providers-http            2.0.2
apache-airflow-providers-imap            2.2.3
apache-airflow-providers-postgres        4.1.0
apache-airflow-providers-redis           2.0.4
apache-airflow-providers-sendgrid        2.0.4
apache-airflow-providers-sftp            2.6.0
apache-airflow-providers-slack           4.2.3
apache-airflow-providers-sqlite          2.1.3
apache-airflow-providers-ssh             2.4.3

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

Here are some logs from scheduler pod, found kubernetes_executor try to delete worker pod.
```bash
[2022-05-13 02:48:08,417] {kubernetes_executor.py:147} INFO - Event: testdag2task1.34ae944f04ee4a18a65f559163ca0d1a had an event of type MODIFIED
[2022-05-13 02:48:08,417] {kubernetes_executor.py:206} INFO - Event: testdag2task1.34ae944f04ee4a18a65f559163ca0d1a Succeeded
[2022-05-13 02:48:08,423] {kubernetes_executor.py:147} INFO - Event: testdag2task1.34ae944f04ee4a18a65f559163ca0d1a had an event of type DELETED
[2022-05-13 02:48:08,423] {kubernetes_executor.py:206} INFO - Event: testdag2task1.34ae944f04ee4a18a65f559163ca0d1a Succeeded
[2022-05-13 02:48:08,497] {backfill_job.py:397} INFO - [backfill progress] | finished run 0 of 25 | tasks waiting: 5 | succeeded: 0 | running: 5| failed: 0 | skipped: 0 | deadlocked: 0 | not ready: 5
[2022-05-13 02:48:13,390] {kubernetes_executor.py:375} INFO - Attempting to finish pod; pod_id: testdag2task1.34ae944f04ee4a18a65f559163ca0d1a; state: None; annotations: {'dag_id': 'test_dag_2', 'task_id': 'task_1', 'execution_date': None, 'run_id': 'backfill__2022-03-03T02:00:00+00:00', 'try_number': '1'}
[2022-05-13 02:48:13,391] {kubernetes_executor.py:375} INFO - Attempting to finish pod; pod_id: testdag2task1.34ae944f04ee4a18a65f559163ca0d1a; state: None; annotations: {'dag_id': 'test_dag_2', 'task_id': 'task_1', 'execution_date': None, 'run_id': 'backfill__2022-03-03T02:00:00+00:00', 'try_number': '1'}
[2022-05-13 02:48:13,392] {kubernetes_executor.py:576} INFO - Changing state of (TaskInstanceKey(dag_id='test_dag_2', task_id='task_1', run_id='backfill__2022-03-03T02:00:00+00:00', try_number=1), None, 'testdag2task1.34ae944f04ee4a18a65f559163ca0d1a', 'jutopia-chlee-test-chlee-backfill-test', '3559062271') to None
[2022-05-13 02:48:13,396] {kubernetes_executor.py:661} INFO - Deleted pod: TaskInstanceKey(dag_id='test_dag_2', task_id='task_1', run_id='backfill__2022-03-03T02:00:00+00:00', try_number=1) in namespace jutopia-chlee-test-chlee-backfill-test
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by snjypl**
the backfill queued tasks were getting reset to scheduled state by `kubernetes_executer.clear_not_launched_queued_tasks` of   `scheduler` instance. 

this fixes the issue by making `clear_not_launched_queued_tasks` only handle the tasks which are created  by its own `scheduler/job`.  

closes: #23356 , closes: #23693 , closes:  #23145

### CODE REVIEW

**Comment by potiuk** [comment]
@ashb @ephraimbuddy @dstandish @jedcunningham -> not sure if that's the right fix, But I guess this addresses the issues for 2.3.1

**Comment by ei-grad** [comment]
Having the same issue in airflow==2.2.5. Applying a slightly adapted patch from this pull request fixes the problem with broken backfill.

File: airflow/executors/kubernetes_executor.py
Comment (uranusjr) [inline airflow/executors/kubernetes_executor.py]: Any reason why this excludes `manual`?

Comment (snjypl) [reply airflow/executors/kubernetes_executor.py]: to me it seemed like, for `manual` the `ti.queued_by_job_id` is not updated, it remains same as the original job_id. 

and also, in this case the executor.job_id is str, and ti.queued_by_job_id is expected to the an integer.  so the query in `clear_not_launched_queued_tasks` will fail.

**Comment by vchiapaikeo** [comment]
Hi team! New to the scene here, but we are also facing this bug as well in our implementation (2.3.3, KubernetesExecutor). Wondering if this PR could get a second pair of eyes. Similar to what is mentioned in the discussion, we'll have backfills that run (some to completion) but then get stuck because they are later marked `scheduled`.

All tasks getting marked as scheduled after already completing their backfill:
image

A dag run that completes one task but has its subsequent task marked as scheduled:
image

**Comment by ephraimbuddy** [comment]
> I don't think we should change how Kubernetes executor handles queued task instances that have overstayed in queued state. If a task stayed in queued state for a long time, then something is wrong and the task should be cleared since we have a configuration to detect and clear it. While #26205 can also solve this issue, I think it's not necessary to send a task that's not in `scheduled` state to the executor. This is missing right now in backfill because the possibility to have many tasks set to scheduled was solved in #26161.

After consultations, I'm withdrawing the above statement.

File: tests/executors/test_kubernetes_executor.py
Comment (ephraimbuddy) [inline tests/executors/test_kubernetes_executor.py:163]: What's the reason for changing this to 1?

**Comment by ephraimbuddy** [comment]
Can you add tests to ensure this is not altered in the future?

**Comment by agomez-etsy** [comment]
Hi all. I can confirm I still see this issue on 2.4.0 KubernetesExecutor. Would this be available in 2.4.2? I see the milestone has being pushed several times. We are extremely interested in this fix

**Comment by ephraimbuddy** [comment]
> Hi all. I can confirm I still see this issue on 2.4.0 KubernetesExecutor. Would this be available in 2.4.2? I see the milestone has being pushed several times. We are extremely interested in this fix

We need tests on this one

**Comment by vchiapaikeo** [comment]
For what it's worth, we just updated to 2.4.1 yesterday (w/ KubernetesExecutor) and we are unable to repro the previous buggy behavior where the scheduler would change the status of backfill runs. @ephraimbuddy , I think your PR may have mitigated or resolved this issue -https://github.com/apache/airflow/pull/26342/files.

By fixed, I mean that we can now start a backfill with the `airflow dags backfill` command and have it run to completion. It no longer seems to be getting interrupted mid-way by a scheduler process that is resetting its state.

image

**Comment by vchiapaikeo** [comment]
Argh, I may have spoken too soon. Another stuck one 😢 

image

**Comment by vchiapaikeo** [comment]
One more note - we see these log lines directly before the backfill stalls. Which I think is in line with what this PR will address since @snjypl is modifying the function `clear_not_launched_queued_tasks`.

```
[2022-10-04 18:37:06,838] {scheduler_job.py:1374} INFO - Resetting orphaned tasks for active dag runs
--
  |   | [2022-10-04 18:37:08,918] {kubernetes_executor.py:470} INFO - Found 2 queued task instances
  |   | [2022-10-04 18:37:08,930] {kubernetes_executor.py:511} INFO - TaskInstance: <TaskInstance: test_spark_yaml_1664908327.dataproc_cluster.check_failed_jobs backfill__2022-09-06T00:00:00+00:00 [queued]> found in queued state but was not launched, rescheduling
  |   | [2022-10-04 18:37:08,944] {kubernetes_executor.py:511} INFO - TaskInstance: <TaskInstance: test_spark_yaml_1664908327.dataproc_cluster2.check_failed_jobs backfill__2022-09-06T00:00:00+00:00 [queued]> found in queued state but was not launched, rescheduling
```

**Comment by snjypl** [comment]
> Can you add tests to ensure this is not altered in the future?

@ephraimbuddy  , i have added some tests, can you please review it and let me know if i needto change/update anything.  thank you for help !

File: airflow/executors/kubernetes_executor.py
Comment (ephraimbuddy) [inline airflow/executors/kubernetes_executor.py:472]: Sorry for taking so long to come to this PR, I was on vacation. 

If I understand correctly, this is the main change addressing the issue in this PR. Can we have only this change plus the test on this PR? All other changes should be moved to a new PR. This helps us track changes and it's more specific to the issue being resolved too

Comment (ei-grad) [reply airflow/executors/kubernetes_executor.py:472]: > Can we have only this change plus the test on this PR?

Wouldn't `self.job_id` have value `"backfill"` without the change in [airflow/jobs/backfill_job.py](https://github.com/apache/airflow/pull/23720/files#diff-ce6cb8661a9f4cb5453effb446e6fbccf42a61d597445fe98899051db61436db)?

Comment (ephraimbuddy) [reply airflow/executors/kubernetes_executor.py:472]: Should have it

**Comment by potiuk** [comment]
Needs conflict resolving.

File: airflow/executors/kubernetes_executor.py
Comment (ei-grad) [reply airflow/executors/kubernetes_executor.py:472]: Would it be possible to have more than one backfill running then?

**Comment by potiuk** [comment]
Some strange errors to fix.

**Comment by ei-grad** [comment]
Looks like the only meaningful error is:
```
Run black (python formatter) on other..................................................Failed
- hook id: black
- duration: 67.5s
- files were modified by this hook

reformatted tests/executors/test_kubernetes_executor.py
```

**Comment by potiuk** [comment]
Yep. You need to fix it - I recommend installing pre-commit (as per our contribution docs). It will fix it for you.

**Comment by potiuk** [comment]
Make sure to rebase BTW.

**Comment by snjypl** [comment]
> Yep. You need to fix it - I recommend installing pre-commit (as per our contribution docs). It will fix it for you.

@potiuk i have fixed the issue with black formatter by installing pre commit hook. 
there is still a failing task. 
Tests / Sqlite Py3.7: API Always CLI Core Integration Other Providers WWW 

does it need to be addressed?

**Comment by DMilmont** [comment]
Is there anything I can do to help move this one along? We rely on the airflow backfill CLI and this bug really makes it difficult for our users to backfill anything. 

Anyone have any workarounds that I can implement now without waiting for this to be merged in?

**Comment by potiuk** [comment]
> Is there anything I can do to help move this one along? We rely on the airflow backfill CLI and this bug really makes it difficult for our users to backfill anything.

You can patch your airflow with those changes and verify that they work in your case. That would be an absolutely awesome input and one that can speed up the merge process if we have some evidences that people are eager to get it, even to the extend that they decide to patch their airflow without waiting for the next release.

The change looks like easy to patch manually.

File: airflow/executors/kubernetes_executor.py
Comment (ephraimbuddy) [reply airflow/executors/kubernetes_executor.py]: @snjypl personally, I'll prefer to have this part in a separate PR because it seems to be addressing something else?

**Review by ephraimbuddy — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by DMilmont** [comment]
I went ahead and manually patched this proposed fix into our airflow environment. 

We average around 30 backfills per day, and over the last 4 days since it was manually patched - I have not seen or heard of this issue popping up. So far it seems that the fix has resolved the issue.

If it pops up again before this is merged I will report back. Just wanted to update with some test results.

**Comment by potiuk** [comment]
> If it pops up again before this is merged I will report back. Just wanted to update with some test results.

Cool. Thanks for that !

**Comment by potiuk** [comment]
Thanks @snjypl

**Pull request merged**
